### PR TITLE
Added tests for Insights popup

### PR DESCRIPTION
**Pull request by quarckster**
This PR adds protractor tests for Insights popup from https://github.com/openshift/console/pull/6660. Executed locally against cluster-bot's cluster:
```
root@thinkpad-t480s:/mnt/frontend# NO_HEADLESS=true BRIDGE_BASE_ADDRESS=url BRIDGE_KUBEADMIN_PASSWORD=password yarn test-protractor-suite --suite=dashboards
yarn run v1.21.0
$ yarn ts-node ./node_modules/.bin/protractor integration-tests/protractor.conf.ts --suite=dashboards
$ ts-node -O '{"module":"commonjs"}' ./node_modules/.bin/protractor integration-tests/protractor.conf.ts --suite=dashboards
Report destination:   gui_test_screenshots/protractor-report.html
[18:01:42] I/launcher - Running 1 instances of WebDriver
[18:01:42] I/hosted - Using the selenium server at http://192.168.1.6:4444/wd/hub

>> Executing 26 defined specs...

Test Suites & Specs:

1) Auth test

   2) Login test
      ✔ logs in as kubeadmin user

3) Create a test namespace
   ✔ creates test namespace if necessary

4) Cluster Dashboard

   5) Details Card
      ✔ has all fields populated
      ✔ has View settings link

   6) Status Card
      ✔ has View alerts link
      ✔ has health indicators

      7) Insights Popup
         ✔ has title
         ✔ has OCM UI link
         ✔ has chart
         ✔ shows correct number of hits

   8) Inventory Card
      ✔ has all items

   9) Utilization Card
      ✔ has all items
      ✔ has duration dropdown

   10) Activity Card
[18:02:28] W/element - more than one element found for locator By(css selector, [href="/k8s/all-namespaces/events"]) - the first result will be used
[18:02:28] W/element - more than one element found for locator By(css selector, [href="/k8s/all-namespaces/events"]) - the first result will be used
      ✔ has View events link
      ✔ has Pause events button

11) Project Dashboard
   ✔ Dashboard is default details page

   12) Details Card
      ✖ has all fields populated (2 failures)
      • has View all link

   13) Status Card
      • has health indicator

   14) Inventory Card
      • has all items

   15) Utilization Card
      • has all items
      • has duration dropdown

   16) Activity Card
      • has View events link
      • has Pause events button

   17) Launcher Card
      • is displayed when CR exists

   18) Resource Quotas Card
      • shows Resource Quotas

>> Done!
...
Summary:

Suites:  18 of 18
Specs:   17 of 26 (9 disabled)
Expects: 82 (2 failures)
Finished in 50.892 seconds
```

### CODE REVIEW

**Review by bond95 — APPROVED** [review]

File: frontend/integration-tests/views/dashboard.view.ts
Comment (spadgett) [inline frontend/integration-tests/views/dashboard.view.ts:16]: We should add data-test-id attributes where possible for the tests. Avoid complicated xpath selectors.

**Comment by spadgett** [comment]
You'll need to run prettier on the files. Thanks!

File: frontend/integration-tests/views/dashboard.view.ts
Comment (quarckster) [reply frontend/integration-tests/views/dashboard.view.ts:16]: @spadgett  There is no way to do that without changing the application code. By the way, there is a specification for locators in UI testing https://ouia.readthedocs.io/en/latest/. I added support of this in various Patternfly components https://github.com/patternfly/patternfly-react/pulls?q=is%3Apr+is%3Aclosed+ouia

**Review by dtaylor113 — APPROVED** [review]
/lgtm
Please note that new e2e tests should be writting in Cypress:  https://github.com/openshift/console/blob/master/frontend/packages/integration-tests-cypress/README.md

File: frontend/integration-tests/tests/dashboards/cluster-dashboard.scenario.ts
Comment (bond95) [inline frontend/integration-tests/tests/dashboards/cluster-dashboard.scenario.ts:104]: Could it be rewritten to async/await style or not?

**Comment by dtaylor113** [comment]
> @dtaylor113 [you told me](https://coreos.slack.com/archives/C6A3NV5J9/p1605706694224400?thread_ts=1605537794.181700&cid=C6A3NV5J9) that you don't have tests for the dashboard that's why I added Insights tests into protractor based test suite

@quarckster ah, right, thanks for reminding me

**Comment by quarckster** [comment]
@dtaylor113 @spadgett do I need to do something else in this PR in order to get it merged?

**Comment by dtaylor113** [comment]
> @dtaylor113 @spadgett do I need to do something else in this PR in order to get it merged?

@quarckster, needs `lgtm` and `approved` labels from @spadgett, and needs to pass CI

**Comment by quarckster** [comment]
> > @dtaylor113 @spadgett do I need to do something else in this PR in order to get it merged?
> 
> @quarckster, needs `lgtm` and `approved` labels from @spadgett, and needs to pass CI

My PR didn't break the existing tests. Should I retest it until not passed?

File: frontend/integration-tests/views/dashboard.view.ts
Comment (spadgett) [inline frontend/integration-tests/views/dashboard.view.ts:16]: Please avoid complicated xpath selectors in tests. It's much better to add a `data-test` attribute to the element and use that.

**Review by spadgett — COMMENTED** [review]
Note that we're trying to migrate away from Protractor to Cypress, so ideally all new tests are written for Cypress.

File: frontend/integration-tests/views/dashboard.view.ts
Comment (quarckster) [reply frontend/integration-tests/views/dashboard.view.ts:16]: @spadgett there is no way to add such attributes to these elements without patching patternfly-core.

**Comment by quarckster** [comment]
> Note that we're trying to migrate away from Protractor to Cypress, so ideally all new tests are written for Cypress.

@spadgett [I was told](https://coreos.slack.com/archives/C6A3NV5J9/p1605706694224400?thread_ts=1605537794.181700&cid=C6A3NV5J9) that you don't have tests for the dashboard that's why I added Insights tests into protractor based test suite

**Comment by dtaylor113** [comment]
> My PR didn't break the existing tests. Should I retest it until not passed?

@quarckster yes, pls /retest

**Comment by dtaylor113** [comment]
@quarckster think you need to run `frontend>yarn lint --fix` and checkin fixes:
```
yarn run v1.21.0
$ NODE_OPTIONS=--max-old-space-size=4096 yarn eslint .
$ eslint --ext .js,.jsx,.ts,.tsx,.json,.gql,.graphql --color .
�[0m�[0m
�[0m�[4m/go/src/github.com/openshift/console/frontend/integration-tests/tests/dashboards/cluster-dashboard.scenario.ts�[24m�[0m
�[0m   �[2m95:40�[22m  �[31merror�[39m    ["auths"] is better written in dot notation          �[2mdot-notation�[22m�[0m
�[0m   �[2m95:72�[22m  �[31merror�[39m    ["auth"] is better written in dot notation           �[2mdot-notation�[22m�[0m
�[0m  �[2m107:11�[22m  �[33mwarning�[39m  Too many nested callbacks (5). Maximum allowed is 4  �[2mmax-nested-callbacks�[22m�[0m
�[0m  �[2m108:28�[22m  �[33mwarning�[39m  Too many nested callbacks (6). Maximum allowed is 4  �[2mmax-nested-callbacks�[22m�[0m
�[0m  �[2m114:23�[22m  �[33mwarning�[39m  Too many nested callbacks (5). Maximum allowed is 4  �[2mmax-nested-callbacks�[22m�[0m
�[0m  �[2m114:24�[22m  �[31merror�[39m    'e' is defined but never used                        �[2m@typescript-eslint/no-unused-vars�[22m�[0m
�[0m  �[2m114:24�[22m  �[31merror�[39m    'e' is defined but never used                        �[2mno-unused-vars�[22m�[0m
�[0m�[0m
�[0m�[31m�[1m✖ 7 problems (4 errors, 3 warnings)�[22m�[39m�[0m
�[0m�[31m�[1m�[22m�[39m�[31m�[1m  2 errors and 0 warnings potentially fixable with the `--fix` option.�[22m�[39m�[0m
�[0m�[31m�[1m�[22m�[39m�[0m
```

File: frontend/integration-tests/views/dashboard.view.ts
Comment (spadgett) [reply frontend/integration-tests/views/dashboard.view.ts:16]: > @spadgett there is no way to add this attribute to these elements without patching patternfly-core.

Is that true? `headerContent` on the `Popover` takes a DOM node, so we could pass a `span` with a `data-test` attribute.

**Comment by spadgett** [comment]
> @spadgett [I was told](https://coreos.slack.com/archives/C6A3NV5J9/p1605706694224400?thread_ts=1605537794.181700&cid=C6A3NV5J9) that you don't have tests for the dashboard that's why I added Insights tests into protractor based test suite

You'd want to create a new test scenario anyway just for the Insights dashboard. You want the test to live in the package with your plugin instead of being put in the general dashboard test. The recommendation would be to create a new Cypress suite under the Insights package.

**Comment by quarckster** [comment]
> You'd want to create a new test scenario anyway just for the Insights dashboard.
There is no Insights Dashboard. Please check the screenshot. It's just another link in the Status card. This link opens a small popup. Cypress test suite doesn't have tests and views for Cluster Dashboard.
image

**Comment by dtaylor113** [comment]
> You'd want to create a new test scenario anyway just for the Insights dashboard. 
>> There is no Insights Dashboard. .... Cypress test suite doesn't have tests and views for Cluster Dashboard.

I think we are suggesting adding a new cypress test which does view the cluster dashboard.  

Example:  frontend/packages/integration-tests-cypress/tests/app/cluster-dashboard.spec.ts
```
describe('Cluster Dashboard', () => {
  before(() => {
    cy.login();
  });

  afterEach(() => {
    checkErrors();
  });

  after(() => {
    cy.logout();
  });

  it('tests Insights popup', () => {
    cy.visit('/dashboards');
    ...
  });
...
```

**Comment by spadgett** [comment]
> > You'd want to create a new test scenario anyway just for the Insights dashboard.
> 
> There is no Insights Dashboard. Please check the screenshot. It's just another link in the Status card. This link opens a small popup. Cypress test suite doesn't have tests and views for Cluster Dashboard.

I understand, but you still want the tests for the Insights card in your package as a separate suite, not as part of the core cluster dashboard tests. Ideally these would be a new test suite written for Cypress, not Protractor even if there are no current dashboard tests yet in Cypress. Otherwise we're adding tech debt that we'll have to migrate to Cypress later anyway.

**Comment by dtaylor113** [comment]
> I understand, but you still want the tests for the Insights card in your package as a separate suite, not as part of the core cluster dashboard tests.

Hi @spadgett, does Insights have a seperate package in frontend, or could new test suite be located at `frontend/packages/integration-tests-cypress/tests/app/cluster-dashboard.spec.ts`?  And should this new test be run as part of CI?  -thanks

**Comment by spadgett** [comment]
/hold

Holding for the moment since the Protractor tests aren't running properly in CI due to a bug in the test script. See #7446

**Comment by spadgett** [comment]
@dtaylor113 Insights has its own package: https://github.com/openshift/console/tree/master/frontend/packages/insights-plugin

The tests should run as part of CI.

**Review by TheRealJon — COMMENTED** [review]
I agree with @spadgett that this should be a Cypress test in the Insights package. Otherwise, as he said, we will just have to rewrite this in the near future.

**Pull request closed without merging**
